src/server.rs

```diff
@@ -227,855 +725,233 @@ impl ServerStats {
     }
 }
 
-impl<C : CommandCreatorSync + 'static> SccacheServer<C> {
-    /// Create an `SccacheServer` bound to `port`, using `storage` as cache storage.
-    fn new(port: u16, storage: Box<Storage>) -> io::Result<SccacheServer<C>> {
-        let listener = try!(TcpListener::bind(&SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), port))));
-        Ok(SccacheServer {
-            sock: listener,
-            token: Token(1),
-            conns: Slab::new_starting_at(Token(2), 128),
-            storage: Arc::new(storage),
-            stats: ServerStats::default(),
-            shutting_down: false,
-            idle_timeout: DEFAULT_IDLE_TIMEOUT,
-            timeout: None,
-            compilers: HashMap::new(),
-            force_recache: env::var("SCCACHE_RECACHE").is_ok(),
-            creator: C::new(),
+/// tokio-proto protocol implementation for sccache
+struct SccacheProto;
+
+impl<I> ServerProto<I> for SccacheProto
+    where I: Io + 'static,
+{
+    type Request = ClientRequest;
+    type RequestBody = ();
+    type Response = ServerResponse;
+    type ResponseBody = ServerResponse;
+    type Error = io::Error;
+    type Transport = SccacheTransport<I>;
+    type BindTransport = future::FutureResult<Self::Transport, io::Error>;
+
+    fn bind_transport(&self, io: I) -> Self::BindTransport {
+        future::ok(SccacheTransport {
+            inner: io.framed(ProtobufCodec::new()),
         })
     }
+}
 
-    /// Get the port on which this server is listening for connections.
-    #[allow(dead_code)]
-    pub fn port(&self) -> u16 { self.sock.local_addr().unwrap().port() }
+/// Implementation of `Stream + Sink` that tokio-proto is expecting. This takes
+/// a `Framed` instance using `ProtobufCodec` and performs a simple map
+/// operation on the sink/stream halves to translate the protobuf message types
+/// to the `Frame` types that tokio-proto expects.
```

Comment:
Yeah that's *almost* enough for this use case except for the one RPC that sends two responses for one request. Otherwise we could use the non-streaming version of tokio-proto but with that we're forced to use the streaming module which brings in some more complexity :(